sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -648,34 +612,44 @@ message Environment {
   string url = 1;
 }
 
-// Description of a function in a Beam pipeline.
+// A specification of a user defined function.
 //
-// Contains one of _or both of_ a UrnWithParameter specifying the function
-// and the specification for how to execute it against a particular
-// SDK's harness.
 message FunctionSpec {
 
-  // (Optional) An SDK-independent specification of this function.
-  // If present, this must _fully_ specify the function.
+  // (Required) A full specification of this function.
+  //
+  // There are two main styles, which need not be fully distinct
   //
-  // For example the distinguished urn "urn:beam:windowfn:FixedWindows" with
-  // payload `{ duration: n }` fully specifies a windowing function which can
-  // be implemented by the SDK constructing the pipeline, by another SDK (for
-  // language-to-language fusion compatibility) or by the runner directly.
+  // 1. A well-known URN like "urn:beam:windowfn:FixedWindows"
+  //    with an agreed-upon payload (e.g. a duration) that the
+  //    SDK generating this pipeline understands, but is also
+  //    possibly understood by a runnner and/or other SDKs.
+  // 2. A URN that the runner does not understand, such as
+  //    "urn:beam:dofn:javasdk:1.0" that might indicate that
+  //    it is just a serialized Java DoFn from a particular
+  //    version of the Beam Java SDK. The payload will often
+  //    then be an opaque message such as bytes in a
+  //    language-specific serialization format.
+  //
+  // In both cases, the SDK generating the pipeline understands
+  // the data attached to the URN.
   UrnWithParameter spec = 1;
 
-  // (Optional) An SDK-specific specification for how to execute this function,
-  // including a specification of the environment in which the function
-  // can be interpreted and executed.
-  SdkFunctionSpec sdk_fn_spec = 2;
+  // (Required) Reference to the specification of the execution environment
+  // required to invoke this function.
+  string environment_id = 2;
 }
 
 // A URN along with a parameter object whose schema is determined by the
 // URN.
 //
-// The URN will often specify a parametric function or transform such as
-// "Top" or "FixedWindows" while the payload would specify _n_ or
-// _duration_, respectively.
+// This structure is reused in two distinct, but compatible, ways:
+//
+// 1. This can be a specification of the function over PCollections
+//    that a PTransform computes.
+// 2. This can be a specification of a user-defined function, possibly
+//    SDK-specific. (external to this message must be adequate context
+//    to indicate the environment in which the UDF can be understood.
```

Comment:
I did like the example of `Top` as a specified parameterized function - to demonstrate that this may be some composite. Up to you if you think it clutters up the documentation, though.

Also you're missing a closing paren.